src/crimson/os/seastore/segment_manager/block.cc

```diff
@@ -61,23 +62,49 @@ static write_ertr::future<> do_writev(
   // size, we need to rebuild here
   bl.rebuild_aligned(block_size);
 
-  std::vector<iovec> iov;
-  bl.prepare_iov(&iov);
-  return device.dma_write(
-    offset,
-    std::move(iov)
-  ).handle_exception([](auto e) -> write_ertr::future<size_t> {
-      logger().error(
-	"do_writev: dma_write got error {}",
-	e);
-      return crimson::ct_error::input_output_error::make();
-  }).then(
-    [bl=std::move(bl)/* hold the buf until the end of io */](size_t written)
-	       -> write_ertr::future<> {
-    if (written != bl.length()) {
-      return crimson::ct_error::input_output_error::make();
+  return seastar::do_with(
+    std::vector<std::pair<uint64_t, std::vector<iovec>>>(),
+    std::move(bl),
+    [&device, offset] (auto& iovs, auto& bl) {
+    size_t iovs_i = 0;
+    size_t index = 0;
+    uint64_t off = offset;
+    iovs.resize(bl.get_num_buffers() / IOV_MAX + 1);
+    iovs[iovs_i].second.resize(
+      std::min(bl.get_num_buffers() - IOV_MAX * iovs_i, (size_t)IOV_MAX));
+    iovs[iovs_i].first = off;
+    for (auto& bp : bl.buffers()) {
+      iovs[iovs_i].second[index].iov_base = (void*)bp.c_str();
+      iovs[iovs_i].second[index++].iov_len = bp.length();
+      off += bp.length();
+      if (index > IOV_MAX) {
+	iovs_i++;
+	index = 0;
+	iovs[iovs_i].first = off;
+	iovs[iovs_i].second.resize(
+	  std::min(bl.get_num_buffers() - IOV_MAX * iovs_i, (size_t)IOV_MAX));
+      }
```

Comment:
This seems like it should be a helper on bufferlist:

```
std::vector<std::pair<uint64_t, std::vector<iovec>> prepare_iovs() const;
```

Reply:
The uint64_t portion of each pair would simply be an offset from the beginning of the buffer, which you'd then adjust to the base offset to which you are writing.  Otherwise, it would work just as prepare_iov() does now.  Other users are likely to want a way to turn a large bufferlist into a vector of iovs for the same reason.

Reply:
got it:-)